rskj-core/src/main/java/co/rsk/net/AsyncNodeBlockProcessor.java

```diff
@@ -97,18 +99,25 @@ public BlockProcessResult processBlock(@Nullable Peer sender, @Nonnull Block blo
             return ignoreBlockResult(block, start);
         }
 
-        // Check if block is ready for processing - if the block is not too advanced, its parent block is in place etc.
-        boolean readyForProcessing = blockSyncService.preprocessBlock(block, sender, false);
-        if (readyForProcessing) {
-            // Validate block if it can be added to the queue for processing
-            if (isBlockValid(block)) {
-                scheduleForProcessing(new BlockInfo(sender, block), blockNumber, blockHash, peer);
+        // Check if block is ready for processing - if the block is not too advanced, its ancestor blocks are in place etc.
+        List<Block> blocksToConnect = blockSyncService.preprocessBlock(block, sender, false);
+        if (!blocksToConnect.isEmpty()) {
+            // if there's only one block to connect without any ancestors, then schedule its processing ( if the block is valid ofc )
+            if (blocksToConnect.size() == 1 && blocksToConnect.get(0).getHash().equals(block.getHash())) {
+                // Validate block if it can be added to the queue for processing
+                if (isBlockValid(block)) {
+                    scheduleForProcessing(new BlockInfo(sender, block), blockNumber, blockHash, peer);
 
-                return scheduledForProcessingResult(block, start);
+                    return scheduledForProcessingResult(block, start);
+                }
+
+                logger.warn("Invalid block with number {} {} from {} ", blockNumber, blockHash, peer);
+                return invalidBlockResult(block, start);
             }
 
-            logger.warn("Invalid block with number {} {} from {} ", blockNumber, blockHash, peer);
-            return invalidBlockResult(block, start);
+            // if besides the block there are some ancestors, connect them all synchronously
+            Map<Keccak256, ImportResult> connectResult = blockSyncService.connectBlocksAndDescendants(sender, blocksToConnect, false);
+            return BlockProcessResult.connectResult(block, start, connectResult);
         }
```

Comment:
To sum up my comments on this method, I was thinking on sth like this (just refactor, same logic after all):

```
    @Override
    public BlockProcessResult processBlock(@Nullable Peer sender, @Nonnull Block block) {
        final Instant start = Instant.now();

        final long blockNumber = block.getNumber();
        final String blockHash = block.getPrintableHash();
        final String peer = sender != null ? sender.getPeerNodeID().toString() : "N/A";

        // Validate block header first to see if its PoW is valid at all
        if (!isBlockHeaderValid(block)) {
            logger.warn("Invalid block with number {} {} from {} ", blockNumber, blockHash, peer);
            return invalidBlockResult(block, start);
        }

        // Check if block is already in the queue
        if (store.hasBlock(block)) {
            logger.trace("Ignored block with number {} and hash {} from {} as it's already in the queue", blockNumber, blockHash, peer);
            return ignoreBlockResult(block, start);
        }

        // Check if block is ready for processing - if the block is not too advanced, its ancestor blocks are in place etc.
        List<Block> blocksToConnect = blockSyncService.preprocessBlock(block, sender, false);
        if (blocksToConnect.isEmpty()) {
            logger.trace("Ignored block with number {} and hash {} from {} as it's not ready for processing yet", blockNumber, blockHash, peer);
            return ignoreBlockResult(block, start);
        }

        // if besides the block there are some ancestors, connect them all synchronously
        boolean readyToProcess = blocksToConnect.size() == 1 && blocksToConnect.get(0).getHash().equals(block.getHash());
        if (!readyToProcess) {
            Map<Keccak256, ImportResult> connectResult = blockSyncService.connectBlocksAndDescendants(sender, blocksToConnect, false);
            return BlockProcessResult.connectResult(block, start, connectResult);
        }

        if (!isBlockValid(block)) {
            logger.warn("Invalid block with number {} {} from {} ", blockNumber, blockHash, peer);
            return invalidBlockResult(block, start);
        }

        // if there's only one block to connect without any ancestors and it is valid, then schedule its processing
        scheduleForProcessing(new BlockInfo(sender, block), blockNumber, blockHash, peer);
        return scheduledForProcessingResult(block, start);
    }
```